modules/git/url/url.go

```diff
@@ -0,0 +1,71 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package url
+
+import (
+	"fmt"
+	"net/url"
+	"regexp"
+	"strings"
+)
+
+// URL represents a git remote URL
+type URL struct {
+	*url.URL
+	extraMark int // 0 no extra 1 scp 2 file path with no prefix
+}
+
+// String returns the URL's string
+func (u *URL) String() string {
+	switch u.extraMark {
+	case 0:
+		return u.String()
+	case 1:
+		return fmt.Sprintf("%s@%s:%s", u.User.Username(), u.Host, u.Path)
+	case 2:
+		return u.Path
+	default:
+		return ""
+	}
+}
+
+var scpSyntaxRe = regexp.MustCompile(`^([a-zA-Z0-9_]+)@([a-zA-Z0-9._-]+):(.*)$`)
```

Comment:
1. The `name` part might contain dash `-`
2. The `host` part might be IPv6

Reply:
1 resolved
2 I don't really think we can support that.

Reply:
1. it seems `~` or `.` wouldn't appear in the unix names (it doesn't do harm either, so it's fine)
2. reason?

Reply:
removed regex